### ISSUE TITLE
Container Managed Security (embedded tomcat) not working as expected 

### ISSUE THREAD

**Issue by drennane** [grails/grails-core#10851]
### Summary

I am trying to achieve certificate based authentication using the 'CLIENT-CERT' method in Grails 3.3.1 (Spring Boot engine) but when I try and access a protected resource, I am not prompted for a cert.

The intention is for the landing page to be a screen with a (click to) login button which will prompt for a browser stored certificate once clicked.

I have created two skeleton Grails applications in github to illustrate the above. Both applications use Container Managed Security to protect the anything in the "/books/*" directory pattern with one using Grails 2.5.5 (https://github.com/drennane/books-grails2) and the second using Grails 3.3.1 with equivalent configuration in Application.groovy (https://github.com/drennane/books-grails3). Steps to reproduce are as below.

Question posted on Stack Overflow;
https://stackoverflow.com/questions/46049799/certificate-authentication-x509-using-spring-security-in-grails-3-0

### Steps to Reproduce

1. Create necessary keys and key/truststores locally and update config in _Events.groovy (Grails 2) and Application.groovy (Grails 3) and create a p12 certificate using same authority and upload to Firefox (or browser of choice)
2. Run the Grails 2 application and click on the Bookstore controller link on the homepage.
3. User should be prompted to provide a certificate

4. Do the same as above for the Grails 3 application (can use same certs and truststores)
5. Click on the Bookstore controller link on the landing page
6. User is not prompted for a certificate.

I have tried many different ways of presenting the config in Application.groovy but to no avail.

### Expected Behaviour

User should be prompted for a certificate in Grails 3 app 

### Actual Behaviour

No certificate is requested from the app.

### Environment Information

- **Operating System**: macOS Sierra ver 10.12.6
- **Grails Version:** Grails 3.3.1
- **JDK Version:** Java 1.8.0_77
- **Container Version (If Applicable):** 7.0.70

### Example Application

https://github.com/drennane/books-grails2
https://github.com/drennane/books-grails3

**Comment by graemerocher** [comment]
Probably needs to be enabled at the application level. See https://stackoverflow.com/questions/18875356/using-tomcat-basic-auth-with-new-webapplicationinitializer

**Comment by drennane** [comment]
I'm not sure I understand what you are pointing me to. The suggestion involves using a classic web.xml with a ServletContainerInitializer but it's not possible to edit a web.xml in Grails 3.3.1 (at least not according to the documentation) and so I have done it all programatically as is how I expect it should be done;
https://github.com/drennane/books-grails3/blob/master/grails-app/init/books3/clientauth/Application.groovy

Maybe I'm misunderstanding something here?

**Comment by drennane** [comment]
@graemerocher I still have not been able to figure the above issue out. I have created a new story in SO to reflect my issue;
https://stackoverflow.com/questions/49532883/client-cert-based-x509-certificate-authentication-in-grails-3-3
and have also contacted your team at Object Computing for additional support.
If you have any further feedback/direction on my issue I would much appreciated you passing it on.

**Comment by nickipowers** [comment]
@drennane We have received your request and will be in touch with you very soon!

**Comment by drennane** [comment]
Thanks @nickipowers 
It has been suggested through correspondance with a member of your team that this is a Spring Security issue yet I'm only trying to achieve Container Managed Security at present and so there is no involvement from SS. Hoping this makes sense.
Neither example below should be using the SS plugin.

**Comment by alvarosanchez** [comment]
@drennane you are correct, when trying Container Managed Security, there is no need for Spring Security. Sorry for the confusion.

That said, it's also true that the recommended approach for Spring Boot (and hence for Grails 3 too) is to use Spring Security for this, where x509 authentication is known to be working well.

Is there any particular reason to not use Spring Security for this?

**Comment by drennane** [comment]
Thanks @alvarosanchez 

The reason is that I'm not sure if the Spring Security implementation of X509 integration supports my requirements. I have asked this question in Slack but have had no response. 
In essence, I only wish to be prompted for a certificate after reaching a login page and clicking on a button. From what I have learned about the SS implementation, it's either on or off, that is, once a request is made to the web application, a certificate is requested for authentication. 
Please advise on whether you feel it should be possible and then I would be happy to move away from the Container Managed Solution. 

Please see the following for additional background;
https://stackoverflow.com/questions/49532883/client-cert-based-x509-certificate-authentication-in-grails-3-3

**Comment by alvarosanchez** [comment]
This is the default behaviour because Spring Security's `X509AuthenticationFilter` is a "pre-authentication" filter, and those are designed to happen before anything else.

However, Spring Security has also the notion of [filter chains](https://grails-plugins.github.io/grails-spring-security-core/snapshot/index.html#filters), which allows you to configure which filters are applied to which URLs.

In order to make this working with Spring Security, instead of [configuring `useX509 = true`](https://grails-plugins.github.io/grails-spring-security-core/snapshot/index.html#x509) (which will enable it globally as you pointed out), you should [configure the chainMap](https://grails-plugins.github.io/grails-spring-security-core/snapshot/index.html#chainmap) to apply the filter only to the URL pattern you want.

Note that because `useX509 == false`, you need to register yourself the filter and its related beans, [as done here](https://github.com/grails-plugins/grails-spring-security-core/blob/master/plugin/src/main/groovy/grails/plugin/springsecurity/SpringSecurityCoreGrailsPlugin.groovy#L1071-L1097).

**Comment by drennane** [comment]
This is an excellent lead @alvarosanchez. Much appreciated. I will sidestep my Container Managed endeavour for now and see if I can get something working using the filter chains.

**Comment by drennane** [comment]
@alvarosanchez I think I still have an issue with the direction suggested above. Once the SSL connection has been established, by landing on home page, I'm not sure it is possible for Spring Security alone to subsequently prompt for a certificate? I see no way of doing this. For this mechanism to work using SS, I need to configure an SSL enabled endpoint which will proactively request a certificate. SS can then go ahead and read the certification that has been presented by the user. This means I'm back to square one, I need to use a hybrid of Container Managed Security (set client-auth=true and set protected url patterns) and Spring Security to get my desired flow.
Please correct me if I'm wrong.

**Comment by alvarosanchez** [comment]
If I understood you correctly, the landing page will contain a login button. 

If that's the case, the button can point to a URI like `/auth/clientcert`, which you can map in the SS filter chain to have the `X509AuthenticationFilter`. Once the client certificate has been read, validated and its info extracted, SS will populate the principal object and store it in the security context, so the user will be authenticated.

**Comment by drennane** [comment]
You are correct in terms of my requirements. 
So, if I map the URL you suggest to the `X509AuthenticationFilter`, Spring Security will force the browser to prompt for a certificate? I thought it was only the container (tomcat) that could achieve this. 
I already have the code in place to read, validate and populate the principal object. It's literally only the prompting I'm stuck on. 
If you can confirm this, I can feel confident in continuing down this path. 

Thanking you in advance again for your help.

**Comment by alvarosanchez** [comment]
Right, so it appears to be that Spring will only do that if you set the property `server.ssl.client-auth` to either `want` or `need`. By doing so, Spring will configure the embedded container accordingly.

**Comment by alvarosanchez** [comment]
Going back to your first message, what is preventing you from configuring the Tomcat connector + web.xml as you have done in https://github.com/drennane/books-grails2/blob/master/scripts/_Events.groovy?

Anyways, for the scope of this issue report, I don't think this is an issue with Grails itself, so I'm closing it. I'm still happy to try to help you via comments, if I can.

**Comment by drennane** [comment]
Thanks @alvarosanchez, to answer your question above, this is what I was trying to do originally based on my assumption that it was not possible in SS alone. What has triggered this initiative is the need to upgrade to Grails 3 and of course _Events.groovy does not exist in Grails 3, right? So I needed to try and implement the equivalent configuration programmatically in Application.groovy. But it has not worked!!

I don't know if the problem lies with Grails or Spring Boot but neither have been disproven, at least to my knowledge. Can you see my other Grails 3 project where I try and replicate the implementation? Maybe you can spot something that I have done wrong?
Maybe you could also advise as to whether you think I should stop working towards a SS solution and revert to trying to replicate the implementation I have working in Grails 2 by utilising Container Managed Security?

**Comment by alvarosanchez** [comment]
Spring Security can be used or not, but either way, as we've learned through the conversation, it's the container the one that needs to ask the client for a certificate. If SS is also used, then it can read the values from the certificate, build a principal and store it in the security context. This will allow you to use SS infrastructure, like `@Secured`, etc. So using Spring Security is complementary to CMS.

As to what you were trying to do in `Application.groovy`, I think you are missing the `web.xml` part. According to the [Spring Boot documentation](https://docs.spring.io/spring-boot/docs/current/reference/html/howto-traditional-deployment.html), it doesn't seem possible to have an embedded container if you also want to have a `web.xml`, but the following links suggest the contrary:

https://stackoverflow.com/a/37067552
https://stackoverflow.com/a/42285870

Anyway, if the above doesn't work, you can always go with the "traditional deployment" path, by having a Tomcat instance where you configure the connector in their `server.xml` and generating a WAR with a `web.xml` inside.

However, if I were you, I would explore the possibility to use a web server (eg Apache httpd) in front of tomcat, and offload SSL there. You have better flexibility in Apache as you could define something like:

```xml
<Location /auth/clientcert>
  SSLOptions +StdEnvVars +ExportCertData
  SSLVerifyClient require
</Location>
```

Note that you would still have access to the client certificate at the backend level. There, you could use your own code to read its information or leverage Spring Security.

**Comment by drennane** [comment]
My referencing of Spring Security above is literally only in regard to whether or not it was capable of being configured to force the browser to request a certificate but yes, seems we've learned that only the container is capable of this. But other than that, Spring Security is what I use for access control in my application so I will be 100% using it after the auth flow is complete (build a principal etc).

But I'm surprised that the Spring Boot documentation suggests it's not possible configure the embedded container as per my requirements. Can you point out where it says it in the doc you link above? I cannot seem find anything that suggests as much especially seen as Spring appears to expose the necessary APIs to do so and set the very configuration properties that I have done so in _Events.groovy in my Grails 2 project. I wonder why these APIs would be available if they did not work? Am I missing something here?


As I'm sure you can understand, it would represent a real step backwards if our development environment could not closely match our production environment and would prefer not to add an additional component to our stack (Apache httpd) which seems totally unnecessary. I'm also looking to move to a executable jar in production also. So it would be disappointing to learn that this is not possible for me.

In any case, it seems the next step is to try and get this working in Spring Boot. A framework I am not intimately familiar with unfortunately but maybe this will prove to be a useful opportunity to learn.

**Comment by alvarosanchez** [comment]
This [issue](https://github.com/grails/grails-core/issues/10113) might prevent the declarative approach (`web.xml`) to work.

I also had a look at your [programmatic attempt](https://github.com/drennane/books-grails3/blob/master/grails-app/init/books3/clientauth/Application.groovy). I debugged it and seems everything is in place, but I don't why the client authentication isn't triggered.

I would suggest that you try the same code in a Spring Boot Project, to see if it works there and/or there is an issue with Grails or not.

**Comment by drennane** [comment]
Thanks @alvarosanchez, I hope to try this today.

>I also had a look at your programmatic attempt. I debugged it and seems everything is in place, but I don't why the client authentication isn't triggered.

Wondering if this would be reason enough to leave a story open for it, or do you guys consider Spring Boot to be first port of call to fix issues (as it's Grails 3's underlying engine)?

**Comment by alvarosanchez** [comment]
Grails doesn't do anything specifically with Tomcat, and so far there is no evidence of any issue within Grails code. Should you find one, I'm happy to reopen this issue.

**Comment by drennane** [comment]
Well, something does not appear to work as it should, but fair enough. I shall continue my investigation and revert.

**Comment by drennane** [comment]
So by way of an update, I'm sure you'll be interested to know ;-) that I got this working in the end.
Ultimately I needed to go all the back down to vanilla tomcat 8 just to confirm that what I was trying to do a few levels of abstraction away was achievable.
So, after getting it to work in tomcat 8 and then copying this config as best I could in Spring Boot (which worked), I went about applying the same in Grails 3.3.x only to find that Grails does not depend on Spring Boot 2.x as of yet (and probably will not until Q4). I still managed to tweak the classes to versions compatible with Spring Boot 1.5 and so got there in the end. 
I think probably the most important piece of the jigsaw was the addition of a tomcat valve component (using the SSLAuthenticator implementation obviously) which was the only way I could manage to get the browser to prompt for a certificate. This then required me to use my custom realm to retrieve the principal from the certificate (I know of no other way around this at present).
Code is as follows;

```
@Bean
    public EmbeddedServletContainerFactory servletContainer() {
        final TomcatEmbeddedServletContainerFactory tomcat = new TomcatEmbeddedServletContainerFactory();
        tomcat.addContextValves(new SSLAuthenticator());

        tomcat.addContextCustomizers(new TomcatContextCustomizer() {
            @Override
            public void customize(Context ctx) {

                String AUTH_ROLE = "mySecureRole";
                ctx.addSecurityRole(AUTH_ROLE);
                ctx.setRealm(new MySecurityRealm())

                LoginConfig config = new LoginConfig();
                config.setAuthMethod("CLIENT-CERT");
                config.setRealmName("MySecurityRealm");
                ctx.setLoginConfig(config);

                SecurityConstraint constraint = new SecurityConstraint();
                constraint.addAuthRole(AUTH_ROLE);
                SecurityCollection collection = new SecurityCollection();
                collection.addPattern("/secure");
                constraint.addCollection(collection);
                ctx.addConstraint(constraint);
            }
        })

        tomcat.addAdditionalTomcatConnectors(createConnector());
        return tomcat;
    }


    private Connector createConnector() {
        Connector connector = new Connector(TomcatEmbeddedServletContainerFactory.DEFAULT_PROTOCOL);

        connector.setPort(8443);
        connector.setSecure(true);
        connector.setScheme("https");

        Http11NioProtocol proto = (Http11NioProtocol) connector.getProtocolHandler();
        proto.setMinSpareThreads(5);
        proto.setSSLEnabled(true);
        proto.setClientAuth("false");
        proto.setSSLProtocol("all");

        proto.setKeystoreFile("/path/store.jks");
        proto.setKeystorePass("changeit");
        proto.setKeystoreType("JKS");
        proto.setKeyAlias("ssl_server");
        proto.setTruststoreFile("/path/store.jts");
        proto.setTruststoreType("JKS");
        proto.setTruststorePass("changeit");

        proto.setSSLVerifyDepth(2);
        return connector;
    }
```
I'm leaving in the connector details for completeness but of course all of the important stuff is happening in the context customizer.
Now, when I visit this web application I **do not** get prompted for a certificate. This *only* happens when I visit the `/secure` path which is exactly what I required.

**Comment by alvarosanchez** [comment]
@drennane wow, awesome! I'm glad you got it working, and thanks for sharing the final solution. This might be useful to other people.
